service/src/java/org/apache/hive/service/cli/thrift/ThriftHttpServlet.java

```diff
@@ -137,32 +138,47 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
           return;
         }
       }
-      // If the cookie based authentication is already enabled, parse the
-      // request and validate the request cookies.
-      if (isCookieAuthEnabled) {
-        clientUserName = validateCookie(request);
-        requireNewCookie = (clientUserName == null);
-        if (requireNewCookie) {
-          LOG.info("Could not validate cookie sent, will try to generate a new cookie");
-        }
-      }
-      // If the cookie based authentication is not enabled or the request does
-      // not have a valid cookie, use the kerberos or password based authentication
-      // depending on the server setup.
-      if (clientUserName == null) {
-        // For a kerberos setup
-        if (isKerberosAuthMode(authType)) {
-          String delegationToken = request.getHeader(HIVE_DELEGATION_TOKEN_HEADER);
-          // Each http request must have an Authorization header
-          if ((delegationToken != null) && (!delegationToken.isEmpty())) {
-            clientUserName = doTokenAuth(request, response);
-          } else {
-            clientUserName = doKerberosAuth(request);
+
+      clientIpAddress = request.getRemoteAddr();
+      LOG.debug("Client IP Address: " + clientIpAddress);
+      String trustedDomain = HiveConf.getVar(hiveConf, ConfVars.HIVE_SERVER2_TRUST_DOMAIN).trim();
+
+      // Skip authentication if the connection is from the trusted domain
+      if (!trustedDomain.isEmpty() &&
+              PlainSaslHelper.isHostFromTrustedDomain(request.getRemoteHost(), trustedDomain)) {
+        LOG.info("No authentication performed because the connecting host " + request.getRemoteHost() +
```

Comment:
We can only support this for non-kerberos auth mode (password based) and look for "Authorization: Basic" header. Extract the username and discard the password. If cookie comes along with the request, we can use the username from the cookie.

My understanding here is that, a new request comes in with "Authorization: Basic" header, we trust the domain, extract the username from auth header, generate a cookie and respond with cookie. If a new request comes back with the cookie, validate the cookie, extract the user name and we are done.

We should set the expectation from clients here in the config description (whether clients should send basic auth header and that password will be used if not from trusted domain and for trusted domains password will be discarded).

Reply:
Thanks for the detailed explanation.

Here's slight explanation for your "We can only support this for non-kerberos auth mode (password based)".
We can support this independent of the authentication method configured for the HiveServer2 to which the client is connecting. But the connection from a trusted domain should connect as if it's connecting with NOSASL in HTTP mode and NONE authentication in binary mode.

This means that when the HS2 is configured to use kerberos, there is no way that a client can connects from a trusted domain and provides kerberos credentials. It *has to* provide credentials as if it's a password based authentication (but password will not be used if provided). Any connection from non-trusted domain should provide kerberos credentials. I think this slight asymmetry is fine for our use case.

When HS2 is configured to use password based authentication, there's no difference between a connection from a trusted domain and a connection from non-trusted domain.

Rest of your explanation makes sense. I have changed the code accordingly.

Please let me know if we need any changes to the config description.